### PR TITLE
Add initial documentation

### PR DESCRIPTION
**Pull request by arporter**
Brings in the documentation from PSyclone (#14).

### CODE REVIEW

File: doc/api.rst [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+The constructor takes two arguments:
+
+ 1. The grid on which the field exists
+ 2. The type of grid point at which the field is defined
+    (``GO_U_POINTS``, ``GO_V_POINTS``, ``GO_T_POINTS`` or ``GO_F_POINTS``)
+
+Note that the grid object need not have been fully configured (by a
+call to ``grid_init`` for instance) before it is passed into this
+constructor.
```
Comment (hiker) [inline doc/api.rst:134]: I don't think this statement is true: the r2d_field_constructor accesses grid%nx, which as far as I can tell is only initialised after grid_init is called:
```
    upper_x_bound = self%grid%nx
    upper_y_bound = self%grid%ny
```
(lines 226 of field_mod.f90)

**Comment by hiker** [comment]
The documentation builds fine, see my comments about minor fixes. I am not entirely sure what the aim of this library is: it is only to be used for PSyclone, or potentially as a standalone library for applications? In the latter case we would at the very least need some examples on how to access field elements, and how to write loops.

Considering that this is an 'initial documentation' pull request, I am happy to accept this without the loop/field access details, but then will open a ticket to add this later (unless it is only a PSyclone infrastructure library).

File: doc/api.rst [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+The infrastructure requires this information in order to determine the
+extent of the model grid.
+
+The index offset is required because a model (kernel) developer has
+choice in how they actually implement the staggering of variables on a
+grid. This comes down to a choice of which grid points in the vicinity
+of a given T point have the same array (*i*, *j*) indices. In
+the diagram below, the image on the left corresponds to choosing those
+points to the South and West of a T point to have the same (*i*, *j*)
+index. That on the right corresponds to choosing those points to the
```
Comment (hiker) [inline doc/api.rst:63]: "That on" sounds odd to me: "The one" ?? I would have just changed it myself, but thought I'll better ask you for things regarding English language :)

**Comment by hiker** [comment]
All is fine except for the two mentioned things!

File: doc/api.rst (hunk 1 above)
Comment (arporter) [reply doc/api.rst:134]: Oh yes, that's true. I've altered the text accordingly.

File: doc/api.rst (hunk 2 above)
Comment (arporter) [reply doc/api.rst:63]: :-) "That on the right" means "the image on the right" and follows from the previous sentence that refers to "the image on the left". I *think* it's correct but I could be wrong...

**Comment by arporter** [comment]
Thanks for the review. I agree that the "documentation" leaves a lot to be desired but I thought I would at least move it out of PSyclone. In principle this is a stand-alone library but in practice it is only used by PSyclone'd applications. We've contemplated switching to something like Atlas (from ECMWF) but have yet to really have a need.

**Comment by arporter** [comment]
Ready for second review from @hiker.

File: doc/api.rst (hunk 2 above)
Comment (hiker) [reply doc/api.rst:63]: Who am I to argue English with a native speaker ;)

File: doc/api.rst (hunk 1 above)
Comment (hiker) [reply doc/api.rst:134]: Great!

**Review by hiker — APPROVED** [review]

**Comment by hiker** [comment]
I had to solve a minor conflict in README.md, wasn't worth going back to you for that :)

**Pull request merged**